File: python/algos/sorted_array_to_bst.py
from __future__ import annotations

from typing import Sequence


class TreeNode:
    def __init__(self, value: int = 0, left: TreeNode | None = None, right: TreeNode | None = None) -> None:
        self.val = value
        self.left = left
        self.right = right


def sorted_array_to_bst_ranged(values: Sequence[int], left: int, right: int) -> TreeNode | None:
    if left > right:
        return None
    elif left == right:
        return TreeNode(values[left])
    else:
        middle = left + (right - left) // 2
        return TreeNode(
            values[middle],
            sorted_array_to_bst_ranged(values, left, middle - 1),
            sorted_array_to_bst_ranged(values, middle + 1, right),
        )


def sorted_array_to_bst(values: Sequence[int]) -> TreeNode | None:
    return sorted_array_to_bst_ranged(values, 0, len(values) - 1)


if __name__ == '__main__':
    sorted_array_to_bst([])
    sorted_array_to_bst([1])
    sorted_array_to_bst([-10, -3, 0, 5, 9])

File: rust/src/algos/sorted_array_to_bst.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub fn sorted_array_to_bst_ranged(
    nums: &Vec<i32>,
    left: usize,
    right: usize,
) -> Option<Rc<RefCell<TreeNode>>> {
    return if left > right {
        None
    } else if left == right {
        Option::from(Rc::new(RefCell::new(TreeNode {
            val: nums[left],
            left: None,
            right: None,
        })))
    } else if right - left == 1 {
        Option::from(Rc::new(RefCell::new(TreeNode {
            val: nums[right],
            left: Option::from(Rc::new(RefCell::new(TreeNode {
                val: nums[left],
                left: None,
                right: None,
            }))),
            right: None,
        })))
    } else if right - left == 2 {
        Option::from(Rc::new(RefCell::new(TreeNode {
            val: nums[left + 1],
            left: Option::from(Rc::new(RefCell::new(TreeNode {
                val: nums[left],
                left: None,
                right: None,
            }))),
            right: Option::from(Rc::new(RefCell::new(TreeNode {
                val: nums[right],
                left: None,
                right: None,
            }))),
        })))
    } else {
        let middle = left + (right - left) / 2;
        Option::from(Rc::new(RefCell::new(TreeNode {
            val: nums[middle],
            left: sorted_array_to_bst_ranged(nums, left, middle - 1),
            right: sorted_array_to_bst_ranged(nums, middle + 1, right),
        })))
    };
}

pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    sorted_array_to_bst_ranged(&nums, 0, nums.len() - 1)
}
